azkaban-exec-server/src/main/java/azkaban/dag/DagBuilder.java

```diff
@@ -35,39 +35,83 @@
  */
 public class DagBuilder {
 
-  private final String name;
-  private final DagProcessor dagProcessor;
+  private final Dag dag;
+  private final Map<String, Node> nameToNodeMap = new HashMap<>();
 
-  private final List<NodeBuilder> builders = new ArrayList<>();
-  private final Set<String> nodeNamesSet = new HashSet<>();
+  // The builder can only be used to build a DAG once to prevent modifying an existing DAG after it
+  // is built.
+  private boolean isBuilt = false;
 
+  /**
+   * A builder for building a DAG.
+   *
+   * @param name name of the DAG
+   * @param dagProcessor the associated DagProcessor
+   */
   public DagBuilder(final String name, final DagProcessor dagProcessor) {
     requireNonNull(name, "The name of the DagBuilder can't be null");
-    this.name = name;
     requireNonNull(name, "The dagProcessor of the DagBuilder can't be null");
```

Comment:
Should this be `requireNonNull(dagProcessor, ` ?

Reply:
Fixed. Thanks